cupy/linalg/einsum.py

```diff
@@ -231,119 +371,200 @@ def einsum(*operands):
-
-    if len(converted_inputs) >= 2:
-        results = [view[0] for view in single_views]
-        subscripts = [view[1] for view in single_views]
-        results, subscripts = move_broadcast_axes_to_front(results, subscripts)
-        result, subscript = calc_combined_view(results, subscripts)
-        result, subscript = calc_single_view(result, subscript)
+        if not options['sum_ellipsis']:
+            if '@' not in output_subscript and -1 in dimension_dict:
+                raise ValueError(
+                    "output has more dimensions than subscripts "
+                    "given in einstein sum, but no '...' ellipsis "
+                    "provided to broadcast the extra dimensions.")
+        output_subscript = _parse_ellipsis_subscript(
+            output_subscript, None,
+            ellipsis_len=len(
+                [label for label in dimension_dict.keys() if label < 0])
+        )
+
+        # Make sure output subscripts are in the input
+        tmp_subscripts = set(_concat(input_subscripts))
+        for label in output_subscript:
+            if label not in tmp_subscripts:
+                raise ValueError(
+                    "einstein sum subscripts string included output subscript "
+                    "'%s' which never appeared in an input" % _chr(label))
+        if len(output_subscript) != len(set(output_subscript)):
+            for label in output_subscript:
+                if output_subscript.count(label) >= 2:
+                    raise ValueError(
+                        "einstein sum subscripts string includes output "
+                        "subscript '%s' multiple times" % _chr(label))
+
+    _einsum_diagonals(input_subscripts, operands)
+
+    # no more raises
+
+    if len(operands) >= 2:
+        if any(arr.size == 0 for arr in operands):
+            return cupy.zeros(
+                tuple(dimension_dict[label] for label in output_subscript),
+                dtype=result_dtype
+            )
+
+        # Don't squeeze if unary, because this affects later (in trivial sum)
+        # whether the return is a writeable view.
+        for idx in six.moves.range(len(operands)):
+            arr = operands[idx]
+            if 1 in arr.shape:
+                squeeze_indices = []
+                sub = []
+                for axis, label in enumerate(input_subscripts[idx]):
+                    if arr.shape[axis] == 1:
+                        squeeze_indices.append(axis)
+                    else:
+                        sub.append(label)
+                input_subscripts[idx] = sub
+                operands[idx] = cupy.squeeze(arr, axis=tuple(squeeze_indices))
+                assert len(operands[idx].shape) == len(input_subscripts[idx])
+            del arr
+
+    # unary einsum without summation should return a (writeable) view
+    returns_view = len(operands) == 1
+
+    # unary sum
+    for idx, sub in enumerate(input_subscripts):
+        other_subscripts = list(input_subscripts)
```

Comment:
This `input_subscripts` is guaranteed to be a list object.
Use `copy.copy` instead of `list()`.